### PR TITLE
[Xamarin.Android.Build.Tasks] fix timestamps in various intermediate folders

### PR DESCRIPTION
**Pull request by jonathanpeppers**
When taking an audit of our build times, I was testing the following
scenario:
- `msbuild Droid.csproj /t:Install /bl:first.binlog`
- `msbuild Droid.csproj /t:Install /bl:second.binlog`

In theory, the "second" build should be very fast and basically not do
anything. Unfortunately that was not the case.

In my example, I saw build logs such as:

    Target _LinkAssembliesNoShrink
        Building target "_LinkAssembliesNoShrink" partially, because some output files are out of date with respect to their input files.
        [ResolvedUserAssemblies: Input=C:\Users\myuser\.nuget\packages\xamarin.forms\3.0.0.482510\lib\MonoAndroid10\FormsViewGroup.dll, Output=obj\Debug\android\assets\FormsViewGroup.dll] Input file is newer than output file.
        [ResolvedUserAssemblies: Input=C:\Users\myuser\.nuget\packages\xamarin.forms\3.0.0.482510\lib\MonoAndroid10\Xamarin.Forms.Core.dll, Output=obj\Debug\android\assets\Xamarin.Forms.Core.dll] Input file is newer than output file.
        [ResolvedUserAssemblies: Input=C:\Users\myuser\.nuget\packages\xamarin.forms\3.0.0.482510\lib\MonoAndroid10\Xamarin.Forms.Platform.Android.dll, Output=obj\Debug\android\assets\Xamarin.Forms.Platform.Android.dll] Input file is newer than output file.
        [ResolvedUserAssemblies: Input=C:\Users\myuser\.nuget\packages\xamarin.forms\3.0.0.482510\lib\MonoAndroid10\Xamarin.Forms.Platform.dll, Output=obj\Debug\android\assets\Xamarin.Forms.Platform.dll] Input file is newer than output file.
        [ResolvedUserAssemblies: Input=C:\Users\myuser\.nuget\packages\xamarin.forms\3.0.0.482510\lib\MonoAndroid10\Xamarin.Forms.Xaml.dll, Output=obj\Debug\android\assets\Xamarin.Forms.Xaml.dll] Input file is newer than output file.
        LinkAssemblies

Looking at the `LinkAssemblies` MSBuild task, I didn't see anything
that would be setting the timestamps on copied files.

For timestamps to be correct, we should either:
- Use `MonoAndroidHelper.SetLastAccessAndWriteTimeUtc` in C#
- Use the `<Touch />` MSBuild task after using the `<Copy />` MSBuild
  task

This lead me to write a `CheckTimestamps` unit test that does the
following:
- Store `DateTime.UtcNow` in a variable
- Build an app
- Make sure nothing in `bin` or `obj` are older than the start time

This uncovered even more out of date files such as:
- `mono.android.jar`
- Assemblies in `obj/Debug/linksrc` or `$(MonoAndroidLinkerInputDir)`

Fixes made in various places:
- `<LinkAssemblies />` task needed to use
  `MonoAndroidHelper.SetLastAccessAndWriteTimeUtc` everywhere
  `MonoAndroidHelper.CopyIfChanged` is used
- The `_CopyIntermediateAssemblies` target appeared to have a typo. It
  was running a `<Touch />` on `ResolvedUserAssemblies` where it
  looked like it should be `ResolvedAssemblies` from the above
  `<Copy />`
- The `_AddStaticResources` target needed a `<Touch />` for
  `mono.android.jar`

This fix should improve our incremental build times dramatically, as
these timestamps caused other targets to run that slow things down.

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.cs [hunk 1]
```diff
@@ -120,6 +120,32 @@ public class Foo {
...
+					Directory.Delete (intermediate, true);
+				var output = Path.Combine (Root, b.ProjectDirectory, proj.OutputPath);
+				if (Directory.Exists (output))
+					Directory.Delete (output, true);
+				Assert.IsTrue (b.Build (proj), "Build should have succeeded.");
+
+				//Absolutely non of these files should be *older* than the starting time of this test!
+				var files = Directory.EnumerateFiles (intermediate, "*", SearchOption.AllDirectories).ToList ();
+				files.AddRange (Directory.EnumerateFiles (output, "*", SearchOption.AllDirectories));
+				var now = DateTime.UtcNow;
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.cs]: Whoops I should delete this `now` variable, it's not used.

**Comment by jonathanpeppers** [comment]
My new test failed on macOS, looking into it...

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 2]
```diff
@@ -1730,7 +1731,7 @@ because xbuild doesn't support framework reference assemblies.
 		DestinationFiles="@(_AndroidResolvedSatellitePaths->'$(MonoAndroidLinkerInputDir)%(DestinationSubDirectory)%(Filename)%(Extension)')"
 		SkipUnchangedFiles="true"
 	/>
-	<Touch Files="@(ResolvedUserAssemblies->'$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)')" />
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets]: @dellis1972: Is this change correct? IIRC in Debug builds, `@(ResolvedAssemblies)` may contain *system-installed* assemblies such as `Mono.Android.dll`, which we almost certainly don't want to be `<Touch/>`ing...

Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets]: What made me think this was a mistake was the previous line:
```xml
<Copy
	SourceFiles="@(ResolvedAssemblies)"
	DestinationFiles="@(ResolvedAssemblies->'$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)')"
	SkipUnchangedFiles="true" />
```
I was thinking `DestinationFiles` from `<Copy />` needs to match `Files` from `<Touch />`.

If an item in `@(ResolvedAssemblies)` was the system `Mono.Android.dll`, for example, `@(ResolvedUserAssemblies->'$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)')` should evaluate to `obj\Debug\linksrc\Mono.Android.dll`.

Comment (dellis1972) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets]: I think @jonathanpeppers is right on this one. We are touching the files in the Intermediate dir so it should be ok.

**Comment by dellis1972** [comment]
Nice! `_LinkAssembliesNoShrink` running all the time has bothered me in the past, never had a chance to dig into it though. Good job :)

**Comment by dellis1972** [comment]
@jonathanpeppers might be useful to know we have a `CopyIfChanged` Task [1] as well :)
Similar to `Copy` but does some intelligent stuff. It turns out the `Copy` task checks timestamps and contents and will Copy if either changed. So even if the timestamps match if the contents are different slightly it will copy. In our case we alter some of the intermediate files and even if we reset the timestamps `Copy` will still update them :/ 


[1] https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.Build.Tasks/Tasks/CopyIfChanged.cs

**Comment by dellis1972** [comment]
@jonathanpeppers  @grendello 

`So this looks to be the issue where FileInfo.LastModifiedTimeUtc doesn't have Millisecond information running on Mono. `

Is that a mono bug ? Should it contains millisecond info. Sounds a bit odd that a Unix system would be less accurate than windows?

**Comment by jonathanpeppers** [comment]
> So this looks to be the issue where FileInfo.LastModifiedTimeUtc doesn't have Millisecond information running on Mono.

I saw this behavior when debugging `Xamarin.ProjectTools`, there is also some comments like this: https://github.com/xamarin/xamarin-android/blob/62e7792ca52949f2fa213330d292865ad78d2bfa/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/XamarinProject.cs#L233-L247

**Comment by jonathanpeppers** [comment]
The test failure on Windows, I've seen randomly before:
```
Xamarin.Android.Build.Tests.ResolveSdksTaskTests.ResolveSdkTiming

Task should not take more than 1 second to run.
Expected: less than or equal to 00:00:01
But was: 00:00:01.9140742
```

It worked for me locally, so maybe this should be checking against 2 seconds?

**Comment by dellis1972** [comment]
That test failure is dependent on how loaded the machine is.. :/ I've seen
it take up to 9 seconds on a very busy machine.

**Comment by jonathanpeppers** [comment]
To give an idea of better builds times I'm seeing with this change, our `Xamarin.Forms-Peformance-Integration` project:
- Build, with no changes: 4513ms -> 2772ms
- Build, after touching Android resource file:  19420ms -> 18587ms
- Build, after touching C# file: basically no change

The `HelloWorld` sample did not see much improvement. So I suspect these fixes will present improvements mainly to larger (or Xamarin.Forms) projects.

**Comment by dellis1972** [comment]
@jonathanpeppers looks like you broke windows?

**Comment by jonathanpeppers** [comment]
Yeah it is that 1 test failure for `Xamarin.Android.Build.Tests.ResolveSdksTaskTests.ResolveSdkTiming`.

Maybe `[NonParallelizable]` on this test would help?

**Pull request merged**